### PR TITLE
Fix the tests and use multiple Ruby versions

### PR DESCRIPTION
**Pull request by iHiD**
This PR gets the tests passing again. The track's CI still fails due to configlet failing, but it gives us sanity again for Practice Exercises. I've added inline comments to explain individual changes.

### CODE REVIEW

File: .github/workflows/exercise-tests.yml [hunk 1]
```diff
@@ -19,29 +18,19 @@ jobs:
...
 
     steps:
       - uses: actions/checkout@v2
 
-      - uses: actions/setup-ruby@v1
+      - name: Set up Ruby
+        uses: ruby/setup-ruby@a699edbce608a2c128dedad88e3b6a0e28687b3c
         with:
           ruby-version: ${{ matrix.ruby-version }}
+          bundler-cache: true
```
Comment (iHiD) [inline .github/workflows/exercise-tests.yml:30]: This removes the need for all the code below.

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,5 @@
 .bundle
 bin/configlet
 bin/configlet.exe
-Gemfile.lock
```
Comment (iHiD) [inline .gitignore:4]: I don't know why the `Gemfile.lock` wasn't checked in, but it normally should be, so I've added it.

File: .ruby-version [hunk 3]
```diff
@@ -0,0 +1 @@
+ruby-2.6.6
```
Comment (iHiD) [inline .ruby-version:1]: Ruby 2.7 gives warnings about kwargs, but these warnings are no longer valid as Ruby 3.0 changed how they work, so I've just stuck with 2.6.6 for now, in line with all the other Ruby codebases (e.g. Test Runner)

File: lib/generator/files/track_files.rb [hunk 4]
```diff
@@ -81,7 +81,7 @@ def update_version(version)
 
     class MinitestTestsFile < Writable
       def generate(template:, values:)
-        content = ERB.new(template, nil, '<>').result values.get_binding
+        content = ERB.new(template, trim_mode: '<>').result values.get_binding
```
Comment (iHiD) [inline lib/generator/files/track_files.rb:84]: This is a change to the ERB lib in 2.6 onwards

File: lib/tasks/exercise_test_tasks.rb [hunk 5]
```diff
@@ -16,18 +16,16 @@ def initialize(options:, test_runner: ExerciseTestsRunner)
...
 
-  def exercises
-    @exercises ||= Exercise.all
-  end
-
   def define_task_for_all_exercises
     desc 'Run the tests for all exercises'
-    task exercises: exercises
+    # TODO: Also add concept exercises
+    task exercises: Exercise.practice
```
Comment (iHiD) [inline lib/tasks/exercise_test_tasks.rb:28]: I've tested this with `Exercise.all` and it works in terms of generating the tests correctly, but the concept exercises currently fail the tests due to the structure being different, so I've stuck to practice exercises for now.

File: test/tasks/exercise_test.rb [hunk 6]
```diff
@@ -5,47 +5,61 @@
...
-      end
+      FileUtils.mkdir_p "#{dir}/concept/test"
+      FileUtils.mkdir_p "#{dir}/concept/test2"
+      FileUtils.mkdir_p "#{dir}/practice/test3"
+      FileUtils.mkdir_p "#{dir}/practice/test4"
+
+      concept_file_list = FileList["#{dir}/concept/*"]
+      practice_file_list = FileList["#{dir}/practice/*"]
+      FileList.expects(:[]).with('exercises/concept/*').returns(concept_file_list).at_least_once
+      FileList.expects(:[]).with('exercises/practice/*').returns(practice_file_list).at_least_once
```
Comment (iHiD) [inline test/tasks/exercise_test.rb:16]: I added `mocha` so I can do this ^^

File: lib/tasks/exercise.rb [hunk 7]
```diff
@@ -1,25 +1,36 @@
...
+    def concept_slugs
+      FileList['exercises/concept/*'].pathmap('%f')
+    end
 
-    def slugs
-      FileList['exercises/*'].pathmap('%f').exclude('TRACK_HINTS.md')
+    def practice_slugs
+      FileList['exercises/practice/*'].pathmap('%f')
     end
   end
```
Comment (iHiD) [inline lib/tasks/exercise.rb:22]: All of this needs DRYing up at some point. But let's get everything working first.

**Review by ekingery — APPROVED** [review]

**Comment by neenjaw** [comment]
I don't know how it exactly works all together, but each change looks reasonable with CI passing and listing each exercise.

**Comment by iHiD** [comment]
Thanks both. I'll merge it down and keep going 🙂

**Pull request merged**